src/snapred/backend/data/LocalDataService.py

```diff
@@ -490,12 +486,19 @@ def readNormalizationRecord(self, runId: str, version: str = None):
             latestFile = self._getLatestFile(recordPath)
         record: NormalizationRecord = None  # noqa: F821
         if latestFile:
+            print(f"reading NormalizationRecord from {latestFile}")
             record = parse_file_as(NormalizationRecord, latestFile)  # noqa: F821
+
+            # Read the state's grouping-schema map from its separate JSON file.
+            groupingMap = self._readGroupingMap(stateId)
+            # Attach the grouping-schema map to the normalization record's state:
+            record.normalization.instrumentState.attachGroupingMap(groupingMap)
         return record
 
     def writeNormalizationRecord(self, record: NormalizationRecord, version: int = None):  # noqa: F821
         """
         Persists a `NormalizationRecord` to either a new version folder, or overwrite a specific version.
+        -- side effect: updates version numbers of incoming `NormalizationRecord` and its nested `Normalization`.
```

Comment:
Is this really a side effect?  Regardless I do think its a great idea to explicitly call them out if they are unavoidable.

Reply:
Well, it's a side effect, in the sense that the primary purpose of the method is to _write_ the record.  However, I'd be glad to change the wording.  The only reason I added this comment is that it took me way too long to figure out why the comparison of records at the end of the test was failing (**the "nested" attribute ended up with the incorrect version**).

Reply:
I see, well then experience dictates this is a necessary callout then.  Wording is fine as is.